### DISCUSSION TITLE
Curl leaks .i2p hostnames in DNS

### DISCUSSION THREAD

**Discussion by marek22k** [curl/curl#13964] [Ideas]
Hello,

I recently read [the blog entry](https://daniel.haxx.se/blog/2024/05/17/curl-tor-dot-onion-and-socks/) about curl no longer leaking .onion domains into the DNS, which is great. However, curl still leaks .i2p domains. Like Tor, [I2P](https://geti2p.net/en/) is a "darknet" or overlay network. In contrast to Tor, however, I2P focuses on internal services. In contrast to Tor, I2P is a "small" darknet but still one of the largest. I would therefore also suggest not to leak .i2p domains into the DNS.

There is no corresponding RFC for this - but in my opinion it would still make sense (as the background is the same).

Since I was unsure whether this is a bug, I thought I would start a discussion.

PR: https://github.com/curl/curl/pull/13965

**Comment by dfandrich** [comment]
The .onion case had an RFC behind it, making it easier to justify. Even so, there's been some pushback from a loud contingent about whether curl should be following that RFC. Without an official spec to point to, it will be a harder sell to block .i2p, too.

**Reply by bagder** [reply]
I could only find this abandoned attempt from 2015: https://datatracker.ietf.org/doc/html/draft-grothoff-iesg-special-use-p2p-i2p-00

**Reply by marek22k** [reply]
However, this is more of a meta-argument. The background to .onion, as far as I understand it, is that Tor users who are not very familiar with curl or are not paying attention for a moment communicate the .onion address they want to reach to their DNS provider. This in turn could lead to the DNS provider knowing which hidden services you are on. However, hidden services are designed for anonymity or pseudonymity. In order to prevent such a "leak", this refuse mechanism was implemented.

The same applies to I2P, except that the terminology is different. With I2P it is rather called Eepsite or Destination. However, the actual background is the same.

I think the reason why .onion got an RFC is because Tor is enormously popular and gets support from various organizations. I2P is smaller compared to Tor, but still one of the largest darknets.

**Reply by bagder** [reply]
I think they made an RFC happen for `.onion` because they did the job.

Which authority or register is reserving the `.i2p` TLD?
